File: app/src/main/kotlin/ru/tutu/stations/ui/mvp/presenter/BaseMvpPresenter.java
package ru.tutu.stations.ui.mvp.presenter;

import android.support.annotation.NonNull;
import ru.tutu.stations.ui.mvp.view.MvpView;

/**
 * Скелетная реализация {@link MvpPresenter}.
 *
 * @param <V> {@link MvpView}, c которой работает презентер.
 * @author Grigoriy Pryamov
 */
public abstract class BaseMvpPresenter<V extends MvpView> implements MvpPresenter<V> {

    /**
     * Представление, ассоциированное с презентером.
     */
    protected V view;
    /**
     * Флаг, что презентер проинициализирован, т.е. начал свою работу
     */
    private boolean initialized = false;

    public void initialize() {
        if (!initialized) {
            initialized = true;
            onInitialize();
        }
    }

    abstract void onInitialize();

    public boolean isInitialized() {
        return initialized;
    }

    @Override
    public void bind(@NonNull V view) {
        this.view = view;
    }

    @Override
    public void unbind(@NonNull V view) {
        this.view = null;
    }

    @Override
    public void destroy() {

    }
}

File: app/src/main/kotlin/ru/tutu/stations/ui/mvp/presenter/BaseMvpPresenter.kt
package ru.tutu.stations.ui.mvp.presenter

import ru.tutu.stations.ui.mvp.MvpView

/**
 * Скелетная реализация {@link MvpPresenter}.
 *
 * @author Grigoriy Pryamov.
 */
abstract class BaseMvpPresenter<V : MvpView> : MvpPresenter<V> {
    /**
     * Представление, ассоциированное с презентером.
     */
    protected var view: V? = null
    /**
     * Флаг, что презентер проинициализирован, т.е. начал свою работу
     */
    private var initialized = false

    fun initialize() {
        if (!initialized) {
            initialized = true
            onInitialize()
        }
    }

    protected abstract fun onInitialize()

    override fun bind(view: V) {
        this.view = view
    }

    override fun unbind(view: V) {
        this.view = null
    }

    override fun destroy() {}
}
